### PR TITLE
std.os: add support for socketpair

### LINKED ISSUE #18417: `std.os.socketpair` is not exposed
### Zig Version

0.12.0-dev.1828+225fe6ddb; checked master too

### Steps to Reproduce and Observed Behavior

`std.os.socketpair` is not exposed but the primitives in `os.linux` and `os.darwin` are already there. the constants needed are there as will since `std.os.socket` does exist.

https://man7.org/linux/man-pages/man2/socketpair.2.html
https://man7.org/linux/man-pages/man2/socket.2.html

### Expected Behavior

```zig
const os = std.os;
var fds: [2]c_int = undefined;
try std.os.socketpair(os.AF.UNIX, os.SOCK.STREAM, 0, &fds);
```

that, or return `![2]c_int` and not require the in-parameter

### PR DESCRIPTION
**Pull request by notcancername**
fixes #18417
cc @nektro

with the `AF_UNIX` implementation, the file the socket is bound to is not cleaned up, it might be worth exposing this eventually or abstracting socketpair properly in std.net (related: #6600), comments are appreciated

(also: fix an error where recvfrom would use `std.c` instead of `std.os.windows` even when libc is not linked)

### CODE REVIEW

**Review by nektro — COMMENTED** [review]
some thoughts in no particular order

- nice!
- if windows doesn't have this behavior I don't think std.os should pretend it does imo.
- additionally it should accept `i32` or `c_int` and not bitCast
- `E.OPNOTSUPP, E.PROTONOSUPPORT` shouldn't be in the same case
- neither should `E.NOBUFS, E.NOMEM`
- there shouldn't be util functions, it should all be in the if statement of `socketpair`
- common style for std.os is `if (windows) { _win_ } _posix_`, this is in the wrong order and shouldn't use an `else`

**Comment by notcancername** [comment]
> - if windows doesn't have this behavior I don't think std.os should pretend it does imo.

Seems like `std.net` is the preferred layer for this?

> - `E.OPNOTSUPP, E.PROTONOSUPPORT` shouldn't be in the same case
> - neither should `E.NOBUFS, E.NOMEM`

Which errors would be more appropriate?

> there shouldn't be util functions, it should all be in the if statement of socketpair

They're separate functions because `AF_UNIX` may not be available at runtime:
```zig
        error.AddressFamilyNotSupported => return socketpairEmulatedAfInet(),
```

> additionally it should accept i32 or c_int and not bitCast

`std.c.socketpair`  uses `c_uint`, but `std.os.linux.socketpair` uses `i32`. Some amount of casting is unavoidable with the way it is right now. I've changed `std.os.linux.socketpair` to fit `std.os.socket` precedent.

**Comment by nektro** [comment]
`std.c.socketpair` was wrong. they all should be using `c_int`/`i32`.

https://pubs.opengroup.org/onlinepubs/009696699/functions/socketpair.html
https://www.man7.org/linux/man-pages/man2/socketpair.2.html
https://man.freebsd.org/cgi/man.cgi?query=socketpair(2)

**Comment by nektro** [comment]
```zig
        .OPNOTSUPP => return error.OperationNotSupported,
        .PROTONOSUPPORT => return error.ProtocolNotSupported,
        .NOBUFS => return error.SystemResources,
        .NOMEM => return error.SystemResources,
```

**Comment by notcancername** [comment]
done. why are we using separate switch prongs for the same error?

**Comment by nektro** [comment]
no particular reason afaik. just the style that core team/prs has kept up so far in this case to keep things vertical

**Comment by notcancername** [comment]
these test failures are looking pretty mysterious:

```
 src/Sema.zig:32601:92: error: no field or member function named 'toValue' in 'InternPool.Index'
src/InternPool.zig:2010:19: note: enum declared here
child process failed
Aborted
Error: Process completed with exit code 134.
```

I don't see how my changes could possibly have caused this

**Comment by nektro** [comment]
looks like master branch is failing too

File: lib/std/os/windows/kernel32.zig [hunk 1]
```diff
@@ -452,3 +452,5 @@ pub extern "kernel32" fn RegOpenKeyExW(
 ) callconv(WINAPI) LSTATUS;
 
 pub extern "kernel32" fn GetPhysicallyInstalledSystemMemory(TotalMemoryInKilobytes: *ULONGLONG) BOOL;
+
+pub extern "kernel32" fn GetTempPathA(nBufferLength: DWORD, lpBuffer: LPSTR) callconv(WINAPI) DWORD;
```
Comment (andrewrk) [inline lib/std/os/windows/kernel32.zig:456]: no

**Review by andrewrk — CHANGES_REQUESTED** [review]

**Pull request closed without merging**